### PR TITLE
Implement basic assertions

### PR DESCRIPTION
**Pull request by seb231**

### CODE REVIEW

File: src-cli/witan/models/load_data.clj [hunk 1]
```diff
@@ -196,6 +197,7 @@
   "Input should be a map with keys for each dataset and filepaths to csv
    files as the values. Output is a map of core.matrix datasets."
   [file-map]
+  (utils/property-holds? file-map #(map? %) "Not a map")
```
Comment (acron0) [inline src-cli/witan/models/load_data.clj]: Replace `#(map? %)` with just `map?`

File: src/witan/models/dem/ccm/components_functions.clj [hunk 2]
```diff
@@ -68,6 +86,7 @@
 
 (defn order-ds
   [dataset col-key]
+  (utils/property-holds? dataset #(ds/dataset? %) "Not a dataset")
```
Comment (acron0) [inline src/witan/models/dem/ccm/components_functions.clj]: Replace `#(ds/dataset? %)` with just `ds/dataset?`

File: src/witan/models/dem/ccm/core/projection_loop.clj [hunk 3]
```diff
@@ -18,6 +20,7 @@
    :witan/param-schema {:last-proj-year s/Int}
    :witan/exported? true}
   [{:keys [loop-year]} {:keys [last-proj-year]}]
+  (doseq [y [loop-year last-proj-year]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/core/projection_loop.clj]: As `loop-year` and `last-proj-year` both come directly from the inputs and params, this check _should_ be at the schema level. For example, `:witan/param-schema {:last-proj-year (s/constrained s/Int m-utils/year?)} ...`

File: src/witan/models/dem/ccm/core/projection_loop.clj [hunk 4]
```diff
@@ -49,6 +53,7 @@
                          :population-at-risk PopulationAtRiskSchema}
    :witan/exported? true}
   [{:keys [latest-yr-popn loop-year]} _]
+  (utils/property-holds? loop-year m-utils/year? (str loop-year " is not a year"))
```
Comment (acron0) [inline src/witan/models/dem/ccm/core/projection_loop.clj]: Move check to schema.

File: src/witan/models/dem/ccm/core/projection_loop.clj [hunk 5]
```diff
@@ -80,6 +85,7 @@
    :witan/output-schema {:latest-yr-popn PopulationSchema}
    :witan/exported? true}
   [{:keys [latest-yr-popn births loop-year]} _]
+  (utils/property-holds? loop-year m-utils/year? (str loop-year " is not a year"))
```
Comment (acron0) [inline src/witan/models/dem/ccm/core/projection_loop.clj]: Move check to schema.

File: src/witan/models/dem/ccm/fert/fertility.clj [hunk 6]
```diff
@@ -87,6 +89,7 @@
    {:keys [fert-base-yr]}]
   (let [birth-data-yr (reduce max (ds/column
                                    ons-proj-births-by-age-mother :year))
+        _ (doseq [y [birth-data-yr fert-base-yr]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/fert/fertility.clj]: Move check to schema (fert-base-yr).

File: src/witan/models/dem/ccm/mig/migration.clj [hunk 7]
```diff
@@ -14,6 +16,7 @@
    :witan/output-schema {:projected-domestic-in-migrants ProjDomInSchema}
    :witan/exported? true}
   [{:keys [domestic-in-migrants]} {:keys [start-yr-avg-domin-mig end-yr-avg-domin-mig]}]
+  (doseq [y [start-yr-avg-domin-mig end-yr-avg-domin-mig]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/mig/migration.clj]: Move check to schema.

File: src/witan/models/dem/ccm/mig/migration.clj [hunk 8]
```diff
@@ -26,6 +29,7 @@
    :witan/output-schema {:projected-domestic-out-migrants ProjDomOutSchema}
    :witan/exported? true}
   [{:keys [domestic-out-migrants]} {:keys [start-yr-avg-domout-mig end-yr-avg-domout-mig]}]
+  (doseq [y [start-yr-avg-domout-mig end-yr-avg-domout-mig]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/mig/migration.clj]: Move check to schema.

File: src/witan/models/dem/ccm/mig/migration.clj [hunk 9]
```diff
@@ -38,6 +42,7 @@
    :witan/output-schema {:projected-international-in-migrants ProjInterInSchema}
    :witan/exported? true}
   [{:keys [international-in-migrants]} {:keys [start-yr-avg-intin-mig end-yr-avg-intin-mig]}]
+  (doseq [y [start-yr-avg-intin-mig end-yr-avg-intin-mig]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/mig/migration.clj]: Move check to schema.

File: src/witan/models/dem/ccm/mig/migration.clj [hunk 10]
```diff
@@ -50,6 +55,7 @@
    :witan/output-schema {:projected-international-out-migrants ProjInterOutSchema}
    :witan/exported? true}
   [{:keys [international-out-migrants]} {:keys [start-yr-avg-intout-mig end-yr-avg-intout-mig]}]
+  (doseq [y [start-yr-avg-intout-mig end-yr-avg-intout-mig]] (utils/property-holds? y m-utils/year? (str y " is not a year")) )
```
Comment (acron0) [inline src/witan/models/dem/ccm/mig/migration.clj]: Move check to schema.

File: src/witan/models/dem/ccm/mort/mortality.clj [hunk 11]
```diff
@@ -93,6 +96,8 @@
   [{:keys [historic-asmr future-mortality-trend-assumption]}
    {:keys [start-yr-avg-mort end-yr-avg-mort first-proj-yr
            last-proj-yr mort-scenario mort-last-yr]}]
+  (doseq [y [start-yr-avg-mort end-yr-avg-mort first-proj-yr
```
Comment (acron0) [inline src/witan/models/dem/ccm/mort/mortality.clj]: Move check to schema.

File: src/witan/models/dem/ccm/mort/mortality.clj [hunk 12]
```diff
@@ -131,6 +136,7 @@
    :witan/output-schema  {:deaths DeathsOutputSchema}
    :witan/exported? true}
   [{:keys [initial-projected-mortality-rates population-at-risk loop-year]} _]
+  (utils/property-holds? loop-year m-utils/year? (str loop-year " is not a year"))
```
Comment (acron0) [inline src/witan/models/dem/ccm/mort/mortality.clj]: Move check to schema.

File: src-cli/witan/models/load_data.clj [hunk 13]
```diff
@@ -196,6 +197,7 @@
   "Input should be a map with keys for each dataset and filepaths to csv
    files as the values. Output is a map of core.matrix datasets."
   [file-map]
+  (utils/property-holds? file-map map? "Not a map")
```
Comment (Eleonore9) [inline src-cli/witan/models/load_data.clj:200]: Does the function still run when `file-map` is not a map? 
Should it be inside a `when` block or `try/except` block?

**Comment by acron0** [comment]
:shipit:

File: src-cli/witan/models/load_data.clj (hunk 13 above)
Comment (seb231) [inline src-cli/witan/models/load_data.clj:200]: That's what `property-holds?` fn does. If it's not a map it throws an exception

Comment (acron0) [inline src-cli/witan/models/load_data.clj:200]: Exception will be thrown, terminating the function.

File: src/witan/models/dem/ccm/models_utils.clj [hunk 14]
```diff
@@ -6,6 +6,8 @@
             [witan.models.load-data :as ld]
             [schema.core :as s]))
 
+(defn year? [n] (and (>= n 1900) (<= n 2500)))
```
Comment (Eleonore9) [inline src/witan/models/dem/ccm/models_utils.clj]: Could 2100 be enough as a max year?

File: src/witan/models/dem/ccm/components_functions.clj [hunk 15]
```diff
@@ -16,12 +18,20 @@
   for all the years of interest and is named to avg-name."
   [historical-data col-to-avg avg-name start-yr-avg end-yr-avg]
   (let [hist-earliest-yr (reduce min (ds/column historical-data :year))
-        hist-latest-yr (reduce max (ds/column historical-data :year))
-        start-yr (s/validate (s/pred #(>= % hist-earliest-yr)) start-yr-avg)
-        end-yr (s/validate (s/pred #(<= % hist-latest-yr)) end-yr-avg)]
+        hist-latest-yr (reduce max (ds/column historical-data :year))]
+    (-> start-yr-avg
+        (utils/property-holds?  #(>= % hist-earliest-yr)
```
Comment (Eleonore9) [inline src/witan/models/dem/ccm/components_functions.clj]: It isn't much of a difference, but I wonder if we shouldn't first check if it's a year before checking it's in the right range.

File: src/witan/models/dem/ccm/mig/migration.clj [hunk 16]
```diff
@@ -3,14 +3,16 @@
...
+            [witan.models.dem.ccm.models-utils :as m-utils]
             [clojure.core.matrix.dataset :as ds]
             [schema.core :as s]))
 
 (defworkflowfn project-domestic-in-migrants
   {:witan/name :ccm-mig/proj-dom-in-mig
    :witan/version "1.0.0"
    :witan/input-schema {:domestic-in-migrants DomesticInmigrants}
-   :witan/param-schema {:start-yr-avg-domin-mig s/Int :end-yr-avg-domin-mig s/Int}
+   :witan/param-schema {:start-yr-avg-domin-mig (s/constrained s/Int m-utils/year?) :end-yr-avg-domin-mig (s/constrained s/Int m-utils/year?)}
```
Comment (Eleonore9) [inline src/witan/models/dem/ccm/mig/migration.clj]: Bikeshedding: could you please have the `param/schema` on two lines.

File: src-cli/witan/models/load_data.clj (hunk 13 above)
Comment (Eleonore9) [inline src-cli/witan/models/load_data.clj:200]: ok, thanks!

**Comment by Eleonore9** [comment]
:shipit:

**Pull request merged**